python/lsst/ctrl/bps/wms/panda/panda_service.py

```diff
@@ -132,6 +147,16 @@ def submit(self, workflow):
                 task_cloud=task.cloud,
             )
             idds_client_workflow.add_work(work)
+            if task.is_final:
+                DAG_final_work = work
+            if task.is_dag_end:
+                DAG_end_work.append(work)
+
+        if DAG_final_work:
+            for work in DAG_end_work:
+                cond = Condition(cond=work.is_terminated, current_work=work, true_work=DAG_final_work)
+                idds_client_workflow.add_condition(cond)
+
```

Comment:
Would be helpful to have some comments in these new lines for folks not PanDA experts.  I don't follow how this implements the final job correctly.  My interpretation is that if there's a final job in the GenericWorkflow, for every PanDA task that contains GenericWorkflow sinks ask PanDA (idds?) to run the final job if that task is terminated?  This feels like making the final job into the single sink for the workflow.  (In which case, could have just changed whenMerge to SUCCESS).  If this is temporary code doing exactly what I thought, ignore me.  But if it was supposed to be the final solution, how does only adding this to sink nodes ensure that the final work is done even if a non-sink node fails (i.e., WMS should have pruned the DAG_end_work nodes).

Reply:
I was also going to comment that there is a complete lack of commentary explaining code blocks.